### ISSUE TITLE
feat: add verified iterative self-play policy-improvement loop

### ISSUE THREAD

**Issue by meshwho** [meshwho/PowerGridReconfig#145]
## Goal

Replace the current bulk self-play workflow (`generate a large MCTS dataset -> fine-tune once`) with an autonomous **verified policy-improvement loop**.

The key principle is that MCTS should become a training teacher **only when it proves a better terminal physical result than the current raw neural policy on the same scenario**.

```text
CURRENT CHAMPION θk
        |
        v
   scenario s0
    /       \
RAW NN      MCTS
  |           |
outcome NN  outcome MCTS
  \           /
   improvement gate
        |
   MCTS better?
    /       \
  yes       no
   |         |
store MCTS  discard
trajectory
   |
accepted_count == N
   |
  TRAIN
   |
candidate θk+1
   |
fixed validation + held-out physical evaluation
   |
 candidate better?
   /        \
 yes        no
  |          |
promote     reject
  |          |
 θk+1       θk
   \        /
    next iteration
```

## Why change the current workflow

The existing self-play implementation already provides the important building blocks:

- GraphPolicyValueNetV2 evaluator;
- MCTS with exact AC-PF transitions;
- deterministic scenario-level seeds;
- multiprocessing;
- crash-safe generation resume;
- `ExampleWriter` and current policy/value target semantics;
- training from `--init-checkpoint`;
- scientific evaluation.

The problem is the **selection of training data**. Current self-play stores every completed MCTS episode, even if MCTS does not improve the current network.

The first exploratory run makes this concrete:

- 731 self-play scenarios;
- 2317 generated training states;
- on a fixed unseen hard-200 A/B benchmark, MCTS-300 versus raw NN:
  - MCTS better: 37/200;
  - MCTS worse: 7/200;
  - equal: 156/200;
  - mean final utility improvement: about +0.0093.

So most MCTS episodes do not demonstrate a new policy improvement.

A direct fine-tune on all 731 self-play episodes also shows immediate distribution shift / forgetting on the original validation set:

```text
Epoch 1:
train_policy = 0.799074
val_policy   = 1.551981
val_top1     = 0.5950
val_top5     = 0.8780
val_stop     = 0.7945
val_switch   = 0.5517

Epoch 5:
train_policy = 0.620720
val_policy   = 1.654214
val_top1     = 0.5842
val_top5     = 0.8672
val_stop     = 0.7212
val_switch   = 0.5544
```

Training loss improves while validation degrades, especially the stop/continue decision boundary. This is evidence that unconditional imitation of all MCTS trajectories is not the right self-play update rule.

### 1. Paired RAW-NN / MCTS rollout

For every candidate scenario, start both controllers from the **same initial physical state** and current champion checkpoint:

```text
RAW NN rollout -> terminal physical outcome
MCTS rollout   -> terminal physical outcome
```

RAW NN should be the deterministic policy baseline used for the comparison (greedy legal neural action unless explicitly configured otherwise).

The MCTS trajectory keeps the existing training semantics:

```text
(s_t, pi_MCTS(s_t), U_terminal_MCTS)
```

for every pre-action state in the accepted episode.

Do not change the current value target meaning:

```text
VALUE_TARGET_MODE = final_topology_state_utility
TERMINAL_UTILITY_GAMMA = 1.0
```

### 2. Physical improvement gate

Do **not** accept an episode using only raw scalar objective ordering if doing so would allow a physically worse class.

Use a physical outcome ordering first, e.g.:

```text
physically secure
    > hard-overload-free / safe handoff
    > hard-overloaded
    > PF failure
```

Rules:

```text
if MCTS physical class > RAW physical class:
    ACCEPT

if MCTS physical class < RAW physical class:
    REJECT

if physical class is equal:
    compare terminal J or the monotonic terminal U
```

Within the same physical class, require a meaningful minimum improvement rather than numerical noise, e.g. configurable:

```text
U_MCTS > U_RAW + epsilon_U
```

or an equivalent `J_RAW - J_MCTS > epsilon_J` criterion.

The exact threshold must be explicit, persisted in provenance, and deterministic.

### 3. Collect by accepted improvements, not attempted scenarios

Introduce a target such as:

```text
N = 100 accepted improved episodes
```

Generation continues through unseen scenarios until either:

- `accepted_count == N`, or
- the available scenario pool / configured attempt budget is exhausted.

Do not define one iteration as a fixed number of attempted scenarios.

Persist at least:

```text
attempted
accepted
rejected
acceptance_rate
mean_delta_U among accepted
mean_delta_J among accepted
```

This gives a natural research convergence signal:

```text
p_k = P(MCTS(θ_k) improves RAW(θ_k))
```

and improvement strength:

```text
E[ΔU | accepted]
```

Both should be tracked by iteration.

### 4. Train automatically after N accepted episodes

When `accepted_count == N`:

```text
champion θk
+ accepted MCTS examples
-> existing graph policy/value trainer
-> candidate θk+1
```

Reuse the existing trainer and checkpoint contracts. Do not implement a second training stack.

Normalization must remain compatible with the champion/init checkpoint.

Initially, keep replay simple. The first implementation may train only on the current accepted batch. Replay/teacher anchoring can be added later if empirical forgetting requires it.

### 5. Epoch-0 baseline is mandatory

Before any fine-tuning epoch, evaluate the **unchanged champion checkpoint** on the fixed validation set and record it as `epoch 0`.

Example:

```text
Baseline validation (epoch 0):
val_loss=...
val_policy=...
val_value=...
val_top1=...
val_top3=...
val_top5=...
val_stop=...
val_switch=...
```

`--save-best` must compare fine-tuning epochs against this epoch-0 champion baseline, not only against each other.

If no fine-tuning epoch improves the model-selection criterion relative to epoch 0, the iteration must not silently treat epoch 1 as the best candidate.

### 6. Candidate promotion / rejection

Training completion does **not** automatically promote the checkpoint.

Run candidate versus champion on fixed, unchanged evaluation data.

Promotion must use predeclared criteria based on scientific outcomes, for example:

```text
mean final U:          better / non-inferior
hard-overload-free:    non-inferior
physically secure:     non-inferior
PF failure rate:       non-inferior
```

The existing fixed hard-200 benchmark is suitable for current development experiments, but the implementation should accept an explicit evaluation dataset/config rather than hard-code this artifact.

If the candidate fails promotion:

```text
candidate rejected
champion remains θk
```

The next collection iteration continues from the unchanged champion.

### 7. Autonomous iteration

After promotion/rejection, automatically start the next collection cycle on new scenarios.

Suggested artifact layout:

```text
self_play_loop/
  iter_000/
  iter_001/
      attempts.csv
      accepted_examples.csv
      states/
      candidate.pt
      validation.csv / metrics.csv
      evaluation.json
      manifest.json
  iter_002/
  ...
```

Each iteration must preserve enough provenance for exact reproducibility and crash-safe continuation.

## Reuse and simplify the existing self-play subsystem

Do not build a parallel framework if the existing code can be adapted.

Prefer modifying/reusing the current self-play implementation:

- existing generation runtime;
- evaluator;
- MCTS planner;
- `TopologySwitchingEnv`;
- exact AC PF backend;
- `ExampleWriter`;
- multiprocessing;
- scenario-derived RNG streams;
- progress/resume machinery;
- current training entry point.

The desired change is conceptually:

```text
OLD:
scenario -> MCTS rollout -> always save

NEW:
scenario -> RAW rollout + MCTS rollout -> compare -> save only verified MCTS improvement
```

Then add a thin loop around:

```text
collect N -> train -> validate/evaluate -> promote/reject -> repeat
```

Remove self-play logic that becomes unnecessary under the new deterministic paired-comparison contract rather than keeping obsolete compatibility paths.

## Scientific invariants

This issue must not change:

- optimized AC PF semantics;
- strict Q-limit handling;
- physical state acceptance rules;
- exact topology transition semantics;
- GraphPolicyValueNetV2 architecture semantics;
- policy target definition for accepted MCTS examples;
- terminal topology utility definition;
- `VALUE_TARGET_MODE = final_topology_state_utility`;
- `TERMINAL_UTILITY_GAMMA = 1.0`;
- training objective semantics;
- scientific evaluation metrics.

The key invariant remains:

```text
same scientific input/config/seed/checkpoint
-> same physical problem
-> same accepted PF state
-> same actions/trajectory
-> same targets/tensors
-> same model semantics
-> same MCTS/self-play decisions
-> same scientific evaluation result
```

## Acceptance criteria

- [ ] One scenario can be run as paired RAW-NN and MCTS from the same initial state.
- [ ] Terminal physical outcomes are compared by an explicit deterministic gate.
- [ ] MCTS episodes that are equal or worse are not added to the policy-training buffer.
- [ ] MCTS episodes that are physically better are written in the existing training format.
- [ ] Collection stops at configurable `N` accepted improvements, not a fixed number of attempts.
- [ ] Attempted/accepted/rejected outcomes and `ΔU`/`ΔJ` are persisted.
- [ ] Crash-safe resume does not duplicate attempts or accepted episodes.
- [ ] The current champion is evaluated as validation epoch 0 before fine-tuning.
- [ ] Best-checkpoint selection includes epoch 0 as a valid winner.
- [ ] Candidate promotion is explicit and can reject a trained checkpoint.
- [ ] Rejected candidate leaves the current champion unchanged.
- [ ] Promoted candidate becomes the checkpoint used for the next collection iteration.
- [ ] Iteration-level acceptance rate and accepted-improvement magnitude are recorded.
- [ ] Existing AC-PF / value-target / GraphPolicyValueNetV2 scientific semantics remain unchanged.
- [ ] Tests cover improvement acceptance, tie/worse rejection, physical-class priority, epoch-0 retention, promotion/rejection, and resume behavior.

## Current exploratory data

The existing 731-episode / 2317-state bulk self-play dataset should be treated as an **exploratory iteration-0 artifact**, not as the final autonomous algorithm design. It is useful for diagnostics and comparison but should not constrain the new pipeline to bulk unconditional MCTS imitation.
